### DISCUSSION TITLE
Missing SQLAlchemy documentation references (Sphinx and orm_declarative_metadata)

### DISCUSSION THREAD

**Discussion by unode** [sqlalchemy/sqlalchemy#9998] [Usage Questions] [answered]
I'm not sure where to report this so please apologize if it's not the most adequate place.

When building Sphinx documentation for projects that use SQLAlchemy 2.0+ the following error can be seen:
```python
(...).Base.metadata:6:undefined label: 'orm_declarative_metadata'
```
This occurs because the definition of `orm_declarative_metadata` is in the `doc/` folder which isn't installed with the code.

Usually this isn't much of a concern, but we build our documentation with `sphinx-build -W` which causes warnings to be upgraded to errors. This ultimately leads to sphinx failing to build documentation.

To workaround we are now ignoring sqlalchemy when building docs with:

```python
autodoc_mock_imports = ["sqlalchemy"]
```

The workaround is ok for now but we would like to not have to remove sqlalchemy from the docs entirely.

How are you dealing with this situation in your own projects, and would this qualify as a bug (to be reported)?

**Comment by unode** [comment]
I would but require some guidance on the direction.

Is the solution to remove those references from the docstrings?

I assume that including the source of the doc folder in the package bundle isn't practical.

**Reply by CaselIT** [reply]
I'm not sure how this can be solved.

Have you tried installing sqlalchemy in editable mode when building the docs of your project?

**Reply by unode** [reply]
I have not as we are using sqlalchemy as a dependency and don't clone the repository.

**Comment by zzzeek** [comment]
OK.   Reading this for real now.  I dont understand the issue.

> This occurs because the definition of orm_declarative_metadata is in the doc/ folder which isn't installed with the code.

OK, so source code in our deployed wheel file refers to a label in the documentation, right.

> Usually this isn't much of a concern, but we build our documentation with sphinx-build -W which causes warnings to be upgraded to errors. 

OK....."our" documentation, but....the label you refer to is not in "your" documentation, it's in SQLAlchemy's source code.

How exactly are docstrings inside of SQLAlchemy objects being used in third party documentation?     I would not expect that to work at all without the remote labels failing to resolve.  You would have to tolerate these non-resolved labels or search and replace them using a sphinx extension, possibly with remote-sphinx links (though I've not had good success with sphinx repositories cross-linking each other).

**Comment by unode** [comment]
@zzzeek could you try:

```sh
git clone https://github.com/unode/sqlalchemy-min-repro
cd sqlalchemy-min-repro
./setup-env.sh
./build-docs.sh
```

I believe that what is happening is that the docstring of `DeclarativeBase` is being propagated.

**Reply by zzzeek** [reply]
I'm not sure how Sphinx approaches this issue but I think you would ask on their discussion tracker what's the correct way to deal with this situation.     We have links to our docs in our docstrings, which isn't anything "wrong" on our end.

**Reply by CaselIT** [reply]
the easiest way may be to defined a documentation in your Base so that sphinx does not use the one from the superclass

**Reply by unode** [reply]
At the moment I'm mocking sqlalchemy out, with `autodoc_mock_imports = ["sqlalchemy"]`.
But I'm sure that other people using sphinx to build their documentation will likely run into this problem.
Maybe not critically, if they don't use the `-W` option that turns warnings into errors, but I guess it will depend on how strict they are.

**Reply by unode** [reply]
I tried the docstring override approach but it didn't work because some docstrings are also in attributes (such as `metadata`) which I couldn't override easily.

**Reply by zzzeek** [reply, marked as answer]
you can use [missing-reference](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-missing-reference) hooks to rewrite these references on the fly to either a fixed text or a remote sphinx link

**Reply by unode** [reply]
https://stackoverflow.com/a/62301461 shows how to implement a `missing-reference` handler.
